### PR TITLE
Add a minimal node type for better typescript integration

### PR DESCRIPTION
**Pull request by Martinn1996**
I have added our own node type which contains the minimal properties we use with a generic type. This allows us to use FontoXPath in other TypeScript libraries without having to cast nodes around.

### CODE REVIEW

File: demo/parseAst.ts [hunk 1]
```diff
@@ -1,4 +1,9 @@
 import { IAST } from '../src/parsing/astHelper';
+// import {
```
Comment (DrRataplan) [inline demo/parseAst.ts]: could ya remove the comments?

File: src/index.ts [hunk 2]
```diff
@@ -83,6 +84,8 @@ function compareSpecificity(xpathStringA: string, xpathStringB: string): -1 | 0
  */
 const domFacade = new ExternalDomFacade() as IDomFacade;
 
+declare var window;
```
Comment (JosVerburg) [inline src/index.ts]: Seems work around for strict mode issue. Let's make a proper fix when we enable strict mode.

File: src/expressions/functions/builtInFunctions.fontoxpath.ts [hunk 3]
```diff
@@ -143,14 +143,14 @@ declare const VERSION: string | undefined;
...
-	if (typeof VERSION === 'undefined') {
-		version = 'devbuild';
-	} else {
-		version = VERSION;
-	}
+	version = typeof VERSION === 'undefined' ? 'devbuild' : VERSION;
 	return sequenceFactory.singleton(createAtomicValue(version, 'xs:string'));
 };
 
+// TODO: implement a domparser instead of using the global one from the browser
```
Comment (JosVerburg) [inline src/expressions/functions/builtInFunctions.fontoxpath.ts]: Might be better to create an issue for this than leaving a todo ;)

File: src/parsing/processProlog.ts [hunk 4]
```diff
@@ -261,8 +267,8 @@ export default function processProlog(
 				}${varName.localName} has already been declared.`
 			);
 		}
-		if (!staticContext.lookupVariable(varName.namespaceURI, varName.localName)) {
-			staticContext.registerVariable(varName.namespaceURI, varName.localName);
+		if (!staticContext.lookupVariable(varName.namespaceURI || '', varName.localName)) {
```
Comment (JosVerburg) [inline src/parsing/processProlog.ts]: Why did we add this? Might result in a behavior change.

File: test/specs/parsing/tests/KindTest.tests.ts [hunk 5]
```diff
@@ -46,18 +46,6 @@ describe('KindTest', () => {
 		chai.assert.isTrue(evaluateXPathToBoolean('self::text()', documentNode.documentElement.firstChild));
 	});
 
-	if (typeof DOMParser !== 'undefined') {
```
Comment (JosVerburg) [inline test/specs/parsing/tests/KindTest.tests.ts]: Why was this removed? Was this not used?

File: tsconfig.json [hunk 6]
```diff
@@ -2,11 +2,12 @@
 	"compilerOptions": {
 		"outDir": "./built",
 		"baseUrl": ".",
-		"lib": ["dom", "es6", "es2017", "esnext.asynciterable"],
+		"lib": ["es6", "es2017", "esnext.asynciterable"],
 		"target": "es2017",
 		"module": "commonjs",
 		"inlineSourceMap": true,
 		"declaration": true,
+		"strict": false,
```
[the remark this answers is not in the document]
Comment (Martinn1996) [reply tsconfig.json]: I simply forgot to turn it on again :+1:

File: src/evaluateXPath.ts [hunk 7]
```diff
@@ -159,17 +195,20 @@ export type Options = {
  *
  * @returns The result of executing this XPath
  */
-function evaluateXPath(
+const evaluateXPath = function _evaluateXPath<
```
[the remark this answers is not in the document]
Comment (Martinn1996) [reply src/evaluateXPath.ts]: Otherwise, we couldn't use the generics.

[the remark this answers is not in the document]
Comment (JosVerburg) [reply src/evaluateXPath.ts]: I don't believe that is the case. Something else might have been the problem but I tested it and it does work without the const.

File: src/expressions/dataTypes/Sequences/IteratorBackedSequence.ts [hunk 8]
```diff
@@ -176,13 +176,13 @@ export default class IteratorBackedSequence implements ISequence {
 				if (!isReady) {
 					return notReady(readyPromise);
 				}
-				return mappedResultsIterator.next(IterationHint.NONE);
+				return mappedResultsIterator.next(hint);
```
[the remark this answers is not in the document]
Comment (Martinn1996) [reply src/expressions/dataTypes/Sequences/IteratorBackedSequence.ts]: I think this was a mistake and should indeed have been IterationHint.NONE

File: test/specs/parsing/tests/KindTest.tests.ts (hunk 5 above)
Comment (Martinn1996) [reply test/specs/parsing/tests/KindTest.tests.ts]: The check was not necessary

File: src/index.ts (hunk 2 above)
Comment (DrRataplan) [reply src/index.ts]: No, turns out that lib.dom also defines window. Because we dropped that library, and the window access is optional (part of a UMD wrapper), the declare is needed.

File: test/specs/parsing/tests/KindTest.tests.ts (hunk 5 above)
Comment (DrRataplan) [reply test/specs/parsing/tests/KindTest.tests.ts]: These are the tests, there is never a DOMParser

File: src/expressions/dataTypes/casting/castToHexBinary.ts [hunk 9]
```diff
@@ -12,6 +12,8 @@ function stringToHex(string) {
 	return hex.toUpperCase();
 }
 
+declare var atob;
```
[the remark this answers is not in the document]
Comment (DrRataplan) [reply src/expressions/dataTypes/casting/castToHexBinary.ts]: @Martinn1996 Could you add some comments?

File: src/evaluateXPath.ts [hunk 10]
```diff
@@ -139,8 +140,43 @@ export type Options = {
...
+	[ReturnType.NUMBER]: number;
+	[ReturnType.STRING]: string;
+	[ReturnType.BOOLEAN]: boolean;
+	[ReturnType.NODES]: Node[];
+	[ReturnType.FIRST_NODE]: Node;
+	[ReturnType.STRINGS]: string[];
+	[ReturnType.MAP]: { [s: string]: any };
+	[ReturnType.ARRAY]: any[];
+	[ReturnType.NUMBERS]: number[];
+	[ReturnType.ASYNC_ITERATOR]: AsyncIterator<any>;
```
[the remark this answers is not in the document]
Comment (Martinn1996) [reply src/evaluateXPath.ts]: This should be an AsyncIterator

File: src/expressions/dataTypes/casting/castToBase64Binary.ts [hunk 11]
```diff
@@ -4,13 +4,15 @@ import CastResult from './CastResult';
...
+	let text = '';
 	for (let i = 0; i < hex.length; i += 2) {
-		string += String.fromCharCode(parseInt(hex.substr(i, 2), 16));
+		text += String.fromCharCode(parseInt(hex.substr(i, 2), 16));
 	}
-	return string;
+	return text;
 }
 
+declare var btoa;
```
[the remark this answers is not in the document]
Comment (Martinn1996) [reply src/expressions/dataTypes/casting/castToBase64Binary.ts]: Comments and a type definition has been added.

File: src/expressions/dataTypes/casting/castToHexBinary.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (Martinn1996) [reply src/expressions/dataTypes/casting/castToHexBinary.ts]: Comments and a type definition has been added.

File: src/evaluateXPath.ts [hunk 12]
```diff
@@ -159,17 +195,20 @@ export type Options = {
  *
  * @returns The result of executing this XPath
  */
-function evaluateXPath(
+const evaluateXPath = function _evaluateXPath<
+	T extends Node,
+	TReturnType extends keyof IReturnTypes
```
[the remark this answers is not in the document]
Comment (DrRataplan) [reply src/evaluateXPath.ts]: This is because TypeScript is detecting the type of `evaluateXPath.NODES_TYPE` as `ReturnType`. Declaring the type to be `ReturnType.NODES`, either by casting or declaring it, it does work.

@Martinn1996 if you apply this pattern, it should work:
```
evaluateXPath[
	'BOOLEAN_TYPE'
] = evaluateXPath.BOOLEAN_TYPE = ReturnType.BOOLEAN as ReturnType.BOOLEAN;
```

File: tsconfig.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (Martinn1996) [reply tsconfig.json]: Nevermind, it should stay on false

File: src/index.ts (hunk 2 above)
Comment (JosVerburg) [reply src/index.ts]: Can we add a comment describing the reason we need this? Same as with the other declaration.

File: src/evaluateXPath.ts [hunk 13]
```diff
@@ -159,15 +195,15 @@ export type Options = {
  *
  * @returns The result of executing this XPath
  */
-function evaluateXPath(
+function evaluateXPath<TNode extends Node, TReturnType extends keyof IReturnTypes<TNode>> (
```
Comment (JosVerburg) [inline src/evaluateXPath.ts]: This solution still doesn't seem to work in all cases.
image
However, I don't believe it's worth it to keep this PR open for only those cases. So once the last couple of comments from this review are resolved (don't forget the overall comment which doesn't target a file) I think it's ready for merge.
